clang/lib/AST/Type.cpp

```diff
@@ -4912,3 +4922,279 @@ void AutoType::Profile(llvm::FoldingSetNodeID &ID, const ASTContext &Context) {
   Profile(ID, Context, getDeducedType(), getKeyword(), isDependentType(),
           getTypeConstraintConcept(), getTypeConstraintArguments());
 }
+
+FunctionEffect::~FunctionEffect() = default;
+
+bool FunctionEffect::diagnoseConversion(bool Adding, QualType OldType,
+                                        FunctionEffectSet OldFX,
+                                        QualType NewType,
+                                        FunctionEffectSet NewFX) const {
+  return false;
+}
+
+bool FunctionEffect::diagnoseRedeclaration(bool Adding,
+                                           const FunctionDecl &OldFunction,
+                                           FunctionEffectSet OldFX,
+                                           const FunctionDecl &NewFunction,
+                                           FunctionEffectSet NewFX) const {
+  return false;
+}
+
+bool FunctionEffect::diagnoseMethodOverride(bool Adding,
+                                            const CXXMethodDecl &OldMethod,
+                                            FunctionEffectSet OldFX,
+                                            const CXXMethodDecl &NewMethod,
+                                            FunctionEffectSet NewFX) const {
+  return false;
+}
+
+bool FunctionEffect::canInferOnDecl(const Decl *Caller,
+                                    FunctionEffectSet CallerFX) const {
+  return false;
+}
+
+bool FunctionEffect::diagnoseFunctionCall(bool Direct, const Decl *Caller,
+                                          FunctionEffectSet CallerFX,
+                                          CalleeDeclOrType Callee,
+                                          FunctionEffectSet CalleeFX) const {
+  return false;
+}
+
+const NoLockNoAllocEffect &NoLockNoAllocEffect::nolock_instance() {
+  static NoLockNoAllocEffect global(kNoLockTrue, "nolock");
+  return global;
+}
```

Comment:
I would like it to be this simple, though of the current behaviors of noalloc/nolock, the one that seems extra-special is to detect that a function is declared `nolock(false)` or `noalloc(false)` (by examining the function's type sugar) and short-circuit any attempt to infer the attribute.

Of course that could just be one more bit of behavior, with the identity of the sugar attribute a hardcoded function of the effect type.

Reply:
> detect that a function is declared nolock(false) or noalloc(false) (by examining the function's type sugar) and short-circuit any attempt to infer the attribute.

Well, is that only *functions* or *function types* as well, though? Because functions would still have the `nolock` *attribute*, which then would store that data and could be used for inference. If it’s function types as well, then one option would be to store two bits instead of one, but I agree that that sounds a bit messy and we might want to think a bit more about it in that case. (I’d also have to double-check how/where attributes are stored on types)

Reply:
Since `nolock(true)` needs to be part of a type, it becomes a `FunctionEffect` in a `FunctionEffectSet` attached to a `FunctionProtoType`. I looked at ways to defer `nolock(false)` off of the type and onto the `Decl`, but this violates a hard rule about placement of square-bracket attributes. So `nolock(false)` is implemented as type sugar, via `AttributedType`, even though it's ignored on types and only has meaning when attached to Decls.

Reply:
I may still be misunderstanding something, but by ‘it only has meaning when attached to Decls’, do you mean that e.g.
```c++
using Foo = void () __attribute__((nolock(false)));
__attribute__((nolock)) void foo(Foo f) {
    f(); // Error/warning or sth?
}
```
is something you’re *not* planning to support? In that case, does it even need to appertain to types instead of just to declarations?

Also, the rules for where you can put `__attribute__` and `[[]]` are different in some cases, as I recall, but could you maybe give an example of what exactly you mean by ‘violates a hard rule about placement of square-bracket attributes’ wrt not being able to attach the attribute to the function decl instead of the type here (assuming of course you really only ever want to actually attach it to function declarations, not function types in general).

Reply:
The GNU syntax is more liberal, and yes, this is supported (and tested as of my next push):
```
// Alternate placement on the FunctionDecl
__attribute__((clang_nolock)) void nl_function();
// CHECK: FunctionDecl {{.*}} nl_function 'void () __attribute__((clang_nolock))'
```

One reason I prefer the square-bracket syntax is that it is not only consistent with `noexcept`, but it becomes practical to apply both `nolock` and `noexcept` with a single macro (not sure yet I want to do that to my codebase, but it's possible).

Another is that the placement rules seem more consistent and sensible, to me anyhow.

The `nolock/noalloc(true)` attributes do need to apply to function types, not declarations, so as to be able to analyze indirect calls. They only applies to declarations as a consequence of declarations having function types. `nolock/noalloc(false)` have to be treated the same way as the `true` form for parsing, but for analysis purposes, they end up being ignored until they land on a `Decl`, since indirect calls prevent inference regardless.

As for the rule about square brackets, in SemaType.cpp:
```
    FUNCTION_TYPE_ATTRS_CASELIST:
      attr.setUsedAsTypeAttr();

      // Attributes with standard syntax have strict rules for what they
      // appertain to and hence should not use the "distribution" logic below.
      if (attr.isStandardAttributeSyntax() ||
          attr.isRegularKeywordAttribute()) {
        if (!handleFunctionTypeAttr(state, attr, type, CFT)) {
          diagnoseBadTypeAttribute(state.getSema(), attr, type);
          attr.setInvalid();
        }
        break;
      }
```
So with square-bracket syntax, the attribute has to follow the right parenthesis enclosing the parameter list (and after const).

Incidentally, I realized that in Attr.td the attributes can derive from `TypeAttr` rather than `DeclOrTypeAttr`.

Reply:
I did one whole implementation with both the true/false forms of the attributes as `AttributedType`. Along the way I found and had to hack around two bugs where the sugar gets stripped, e.g. on an `auto` lambda.

I got some guidance that the `true` forms of the attributes made sense as part of the canonical type, so (confusingly, I know):
- the `true` attribute triggers the addition of a global instance of a `NoLockNoAllocEffect` into a `FunctionEffectSet` as an optional part of the `FunctionProtoType`. It does not currently put an attribute on the type.
- the `false` attribute uses `AttributedType`. BTW another limitation of `AttributedType` is that it doesn't seem to hold anything more than the attribute kind, so if we went that way we'd probably need 3 kinds for `nolock` (true, false, type-dependent expression) and 3 for `noalloc`. The type-dependent expression might end up needing a whole new type-sugar class.

Even if `FunctionEffect` turns out to be just some bits controlling details of the behavior, rather than a vtable, it's still attractive to me to encapsulate it into a uniqued object with methods, to centralize all those behaviors.

There's a fair amount of code prepared for the possibility of new, unrelated effects (e.g. "TCB + types"), anticipating that FunctionEffectSet could hold multiple effects. Considering for example a codebase that with several TCB's, they would all have the same behaviors about calls within themselves, but need to be considered unique effects, e.g. a function in TCB A can't call a function that's only part of TCB Z. This is the motivation for the FunctionEffect abstraction.

Reply:
> another limitation of AttributedType is that it doesn't seem to hold anything more than the attribute kind

The attribute itself appears to usually be stored in an `AttributedTypeLoc` rather than as part of the type itself. 

> two bugs where the sugar gets stripped, e.g. on an auto lambda

It seems we’re taking care to try and preserve `AttributedType`s in several places, so that does sound like it might be a bug; we should probably fix those (unless there’s a good reason why we’re stripping the attributes there).

> I got some guidance that the `true` forms of the attributes made sense as part of the canonical type

I agree that that makes sense, yeah.

> Even if FunctionEffect turns out to be just some bits controlling details of the behavior, rather than a vtable, it's still attractive to me to encapsulate it into a uniqued object with methods, to centralize all those behaviors.

I’ll take another look at how it’s currently used; I’d also like to hear what other people (for instance Aaron) have to say about this; I already pinged him in one of the comments below that’s also about this.